sycl/test/check_device_code/device_global_const.cpp

```diff
@@ -0,0 +1,21 @@
+// RUN: %clangxx -fsycl -fsycl-device-only %if cuda %{ -fsycl-targets=nvptx64-nvidia-cuda %} %if hip-amd %{ -fsycl-targets=amdgcn-amd-amdhsa -Xsycl-target-backend --offload-arch=gfx90a %} -S -emit-llvm %s -o - | FileCheck %s  %if cuda || hip-amd %{ --check-prefixes=CHECK-CONST %}
+
+// Tests that device_global<const T> uses const address space for cuda/hip and
+// global address space otherwise
+
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+
+device_global<const int> DeviceGlobalVar;
```

Comment:
I'm not sure this is the right thing.  Why not declare the `device_global` itself as `const` in this case:

```suggestion
const device_global<int> DeviceGlobalVar;
```

It seems like we are moving in a direction where the data type of these "wrapper" classes is expected to be cv-unqualified.  For example, the Khronos WG recently decided that `vec` should only allow a cv-unqualified `DataT`.

Reply:
You could do this, but you would have to make further changes to the specification.

Currently the `memcpy` method of queue/handler does not allow copying a `const device_global` from host to device. This functionalty is absolutely key to making a `const` `device_global` useful.

I did not see any problem with the spec as it is, and the fact that syclomatic is currently already translating codes from `__constant__ to device_global<const T>`, persuaded me that the current solution is the best option. I also think that all things considered, the current solution is the most semantically correct with respect to what the compiler actually does.

If you wish to go down the route of changing the `device_global` spec as described above, that is fine. The only other downside from my perspective is this will slow down blender work a little, and syclomatic will have to be changed. But in the grand scheme of things that doesn't matter I guess.

Reply:
> Currently the `memcpy` method of queue/handler does not allow copying a `const device_global` from host to device. This functionalty is absolutely key to making a `const` `device_global` useful.

This seems OK to me.  In the current specification `const device_global` seems useless.  You can't write the value from either the host or device.  For such a case, why use `device_global` at all?  If you just want a constant, SYCL already allows you to define a constant variable as `const int X = <val>` or `constexpr int X = <val>`.

Reply:
Sorry I should have given more details. In SIMT sometimes values shared across multiple threads in a subgroup will be constant across the threads. In such a case if the GPU retrieves the common value a single time for all threads this will be more efficient than if all threads retrieve the data individially. This is my basic understanding of the motivation for constant caches in GPGPU.

> 
> This seems OK to me. In the current specification `const device_global` seems useless. You can't write the value from either the host or device. For such a case, why use `device_global` at all? If you just want a constant, SYCL already allows you to define a constant variable as `const int X = <val>` or `constexpr int X = <val>`.

I think that in this case this would mean that the program can not update the `const` `X` variable. In such a case I think it would always be more efficient for a programmer to do something like 
```
#DEFINE REG_VAR 10;
```
and then set the value of e.g. a register directly within kernel code to `REG_VAR` without going via any other device memory at all (even .const).

The use case of __constant__ (or in current PR setup here: `device_global<const T>`) is to allow the user to continuously update the constant values at runtime via copying from host (via `queue::memcpy`). This is generally how programs like blender and NWchemex use `__constant__` variables currently with the equivalent cuda API.

Now a third option that I didn't mention previously is to use a attribute like __constant__ (it is a macro but maps to a attribute in attr.td) and not the const keyword. I considered this and got it working, but I also decided against it because I didn't want to complicate the extension by adding a new macro. I then considered getting it to work with the existing "const" *attribute* that is already exposed to all sycl backends in attr.td. Once I realized that `device_global<const T>` works fully within the use cases we need (and that `__constant__` implies `const`) (and also that `const` has the added advantage of already warning users about trying to change the value in kernel code that attributes don't by default), combined with the fact that this "const" attribute in attr.td could be used in different ways in an arbitrary backend that I don't understand, I decided to stick with `device_global<const T>`. Now we could do one of the different options I described and if you tell me to do that I can do it, but be aware that this requires more work, and I do think that the `device_global<const T>` solution that SYCLomatic translates to is a good one.

Reply:
> It seems to me that the cleanest API would be like this:
> 
>     * Recommend that applications use `const device_global<T>` to represent a global variable that is read from device code but never written from device code.
> 
>     * Change SYCLomatic to to migrate to this code pattern.
> 
>     * Change the "copy" member functions that copy to a `device_global` to allow copying even if the `device_global` is declared `const`.  Clarify the wording in the specification to say that it is allowed for the host to write to such `device_global` variables.
> 
> 
> How hard would it be to implement this behavior?

Sure I can change the extension/implementation of device_global to allow this behaviour within this PR. This shouldn't be hard to achieve.

Thanks

Reply:
Sorry to add more complexity here, but I'm not sure I agree with @gmlueck.

There's precedent in SYCL and C++ for these sorts of classes to behave as views, so I don't think it's obvious that a `const device_global` would be one that is read-only on the device.  For example, all `accessor<T>`s are `const` within device code, and one way to represent a read-only `accessor` today is as an `accessor<const T>`.

I think the semantics of a `device_global<const T>` are clearer. It's not `const`, so a user can expect the object to be modified during its execution.  But it provides a constant view.

If we can't reach consensus about this, it might be better to avoid `const` completely, in favor of something like `device_global<T, AccessMode>`, so we can clearly document that the `AccessMode` only applies to device-access, and it can be any of the valid access modes.

Reply:
> If we can't reach consensus about this, it might be better to avoid `const` completely, in favor of something like `device_global<T, AccessMode>`, so we can clearly document that the `AccessMode` only applies to device-access, and it can be any of the valid access modes.

OK, this sounds like a better solution.  The `device_global` class already supports a property named `host_access` which tells whether the variable can be read / written from the host.  Perhaps we could add a property named `device_access` which tells whether the variable can be read / written from the device.

It seems like the only useful access modes are `read_write` (which would be the default) and `read`.  I'm not sure it makes sense to support the other access modes.  In particular, I don't think we structure the API in a way that enforces `write` access mode because there's no way to return a write-only pointer or reference.

Reply:
Are you all happy if I map `device_global<T, AccessMode>` to use .const?
I'm sure I can make it work, although I don't know exactly how to do it immediately and it will probably require me adding an attribute to attr.td.

Thanks

Reply:
Wouldn't the `AccessMode` communicate the same information, just in a less intuitive way? That was part of the accessor simplification for SYCL 2020, to allow `const T` for the data type.

Reply:
When you say `AccessMode`, I assume you really mean that you will add a new compile-time property, right?  I think we should not add a new template parameter `AccessMode`.

I do not think that `const T` is more intuitive.